### ISSUE TITLE
[rush] In 5.57.0-dev.2, "rush change -v" sometimes incorrectly considers ALL projects to be changed

### ISSUE THREAD

**Issue by octogonz** [microsoft/rushstack#3053]
## Summary

While testing **5.57.0-dev.2** in our monorepo, we encountered a regression where `rush change -v` incorrectly detects that every project in the monorepo has been changed.

👉This appears to be a regression introduced by PR #2968.

## Repro steps

The repro steps are in a private monorepo, and I don't have time to prepare them.

## Details

The regression is in this `ProjectChangeAnalyzer.getChangedProjectsAsync()`:

https://github.com/microsoft/rushstack/blob/7d8ff23ab4b500123434e7b5a0ca0400278aece1/apps/rush-lib/src/logic/ProjectChangeAnalyzer.ts#L200-L207

The [getRepoChanges()](https://github.com/microsoft/rushstack/blob/7d8ff23ab4b500123434e7b5a0ca0400278aece1/apps/rush-lib/src/logic/ProjectChangeAnalyzer.ts#L188) call invokes `git diff-index` and correctly identifies the list of files with changes.

However PR #2968 added some extra logic (`TODO: Implement shrinkwrap diffing here`) that basically says _"If the shrinkwrap file has been modified, then assume every project in the monorepo has changes."_

That might make sense for an incremental operation, where processing a few extra projects is okay.  But not for `rush change`.

More fundamentally, I think the code is conflating two subtly different concepts of "impacted by":

1. `rush build`:  Which build tasks **depend upon** files touched by this diff?  ...versus...
2. `rush change`:  Which projects **are parent folders of** source files that are touched by this diff?

The second concept is much more narrow.

## Standard questions

Please answer these questions to help us investigate your issue more quickly:

| Question | Answer |
| -------- | -------- |
| `rushVersion` from rush.json? | 5.57.0-dev.2 |
| `useWorkspaces` from rush.json? | true |

**Comment by octogonz** [comment]
Until this is fixed, HBO will revert our upgrade to **5.57.0-dev.2**.

This should block the release of **5.67.0**.  If it can't be fixed easily, we should consider temporarily reverting PR #2968.

CC @dmichon-msft @iclanton @chengcyber @MickeyPhoenix @dharrie-hbo

**Comment by octogonz** [comment]
For reference, this is the `rush change` callstack that lands in `getChangedProjectsAsync()`:
```
getChangedProjectsAsync
  (.\common\temp\install-run\@microsoft+rush@5.57.0-dev.2\node_modules\@microsoft\rush-lib\lib\logic\ProjectChangeAnalyzer.js:151:62), <anonymous>:1:1)
at ProjectChangeAnalyzer.getChangedProjectsAsync
  (.\common\temp\install-run\@microsoft+rush@5.57.0-dev.2\node_modules\@microsoft\rush-lib\lib\logic\ProjectChangeAnalyzer.js:151:62)
at ChangeAction._getChangedProjectNamesAsync
  (.\common\temp\install-run\@microsoft+rush@5.57.0-dev.2\node_modules\@microsoft\rush-lib\lib\cli\actions\ChangeAction.js:266:61)
at ChangeAction._verifyAsync
  (.\common\temp\install-run\@microsoft+rush@5.57.0-dev.2\node_modules\@microsoft\rush-lib\lib\cli\actions\ChangeAction.js:249:44)
at ChangeAction.runAsync
  (.\common\temp\install-run\@microsoft+rush@5.57.0-dev.2\node_modules\@microsoft\rush-lib\lib\cli\actions\ChangeAction.js:149:24)
at ChangeAction.onExecute
  (.\common\temp\install-run\@microsoft+rush@5.57.0-dev.2\node_modules\@microsoft\rush-lib\lib\cli\actions\BaseRushAction.js:70:21)
at ChangeAction.onExecute
  (.\common\temp\install-run\@microsoft+rush@5.57.0-dev.2\node_modules\@microsoft\rush-lib\lib\cli\actions\BaseRushAction.js:107:22)
at processTicksAndRejections
  (internal/process/task_queues.js:95:5)
at async RushCommandLineParser.onExecute
  (.\common\temp\install-run\@microsoft+rush@5.57.0-dev.2\node_modules\@rushstack\ts-command-line\lib\providers\CommandLineParser.js:199:13)
at async RushCommandLineParser._wrapOnExecuteAsync
  (.\common\temp\install-run\@microsoft+rush@5.57.0-dev.2\node_modules\@microsoft\rush-lib\lib\cli\RushCommandLineParser.js:169:9)
```
